go/tasks/plugins/array/k8s/task.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return Error, err
+	}
+
+	if !succeeded {
+		return Waiting, nil
+	}
+
+	err = kubeClient.GetClient().Create(ctx, pod)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		if k8serrors.IsForbidden(err) {
+			if strings.Contains(err.Error(), "exceeded quota") {
+				// TODO: Quota errors are retried forever, it would be good to have support for backoff strategy.
+				logger.Infof(ctx, "Failed to launch  job, resource quota exceeded. Err: %v", err)
+				t.State = t.State.SetPhase(arrayCore.PhaseWaitingForResources, 0).SetReason("Not enough resources to launch job")
+			} else {
+				t.State = t.State.SetPhase(arrayCore.PhaseRetryableFailure, 0).SetReason("Failed to launch job.")
+			}
+
+			t.State = t.State.SetReason(err.Error())
+			return ReturnState, nil
+		}
+
+		return Error, errors2.Wrapf(ErrSubmitJob, err, "Failed to submit job.")
+	}
+
+	return Success, nil
+}
+
+func (t Task) Monitor(ctx context.Context, tCtx core.TaskExecutionContext, kubeClient core.KubeClient, dataStore *storage.DataStore, outputPrefix, baseOutputDataSandbox storage.DataReference) (TaskStatus, error) {
+	indexStr := strconv.Itoa(t.ChildIdx)
+	podName := formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), indexStr)
+	phaseInfo, err := CheckPodStatus(ctx, kubeClient,
+		k8sTypes.NamespacedName{
+			Name:      podName,
+			Namespace: tCtx.TaskExecutionMetadata().GetNamespace(),
+		})
+	if err != nil {
+		return Error, errors2.Wrapf(ErrCheckPodStatus, err, "Failed to check pod status.")
+	}
+
+	if phaseInfo.Info() != nil {
+		t.LogLinks = append(t.LogLinks, phaseInfo.Info().Logs...)
+	}
+
+	if phaseInfo.Err() != nil {
+		t.MessageCollector.Collect(t.ChildIdx, phaseInfo.Err().String())
+	}
+
+	actualPhase := phaseInfo.Phase()
+	if phaseInfo.Phase().IsSuccess() {
+		originalIdx := arrayCore.CalculateOriginalIndex(t.ChildIdx, t.State.GetIndexesToCache())
+		actualPhase, err = array.CheckTaskOutput(ctx, dataStore, outputPrefix, baseOutputDataSandbox, t.ChildIdx, originalIdx)
+		if err != nil {
+			return Error, err
+		}
+	}
+
+	t.NewArrayStatus.Detailed.SetItem(t.ChildIdx, bitarray.Item(actualPhase))
+	t.NewArrayStatus.Summary.Inc(actualPhase)
+
+	return Success, nil
+}
+
+func (t Task) Abort() {}
+
+func (t Task) Finalize(ctx context.Context, tCtx core.TaskExecutionContext, kubeClient core.KubeClient) error {
+	indexStr := strconv.Itoa(t.ChildIdx)
+	podName := formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), indexStr)
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       PodKind,
+			APIVersion: metav1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      podName,
+			Namespace: tCtx.TaskExecutionMetadata().GetNamespace(),
+		},
+	}
+
+	err := kubeClient.GetClient().Delete(ctx, pod)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
```

Comment:
if the pod doesn't exist, would it make sense to deallocate the resource just as a safeguard below?

Reply:
That makes sense. Updated.

Reply:
I think we attempt to deallocateResources in Finalize. Do we need to do it here again?

Reply:
No we do not. Updated.